src/Database/Table/ColumnAccessCache.php

```diff
@@ -0,0 +1,195 @@
+			$trace[] = isset($item['file'], $item['line']) ? $item['file'] . $item['line'] : null;
+		}
+
+		$key[] = $trace;
+		return $this->generalCacheKey = md5(serialize($key));
+	}
+
+
+	public function setGeneralCacheKey(?string $key): void
+	{
+		$this->generalCacheKey = $key;
+	}
+
+
+	/**
+	 * Returns object specific cache key dependent on query parameters
+	 * Used e.g. for reference memory caching
+	 */
+	public function getSpecificCacheKey(): string
+	{
+		if ($this->specificCacheKey) {
+			return $this->specificCacheKey;
+		}
+
+		return $this->specificCacheKey = $this->selection->getSqlBuilder()->getSelectQueryHash($this->getPreviousAccessedColumns());
+	}
+
+
+	public function setSpecificCacheKey(?string $key): void
+	{
+		$this->specificCacheKey = $key;
+	}
+
+
+	/**
+	 * @return mixed
+	 */
+	public function getAccessedColumns()
+	{
+		return $this->accessedColumns;
+	}
+
+
+	/**
+	 * @param mixed
+	 */
+	public function setAccessedColumns($accessedColumns): void
+	{
+		$this->accessedColumns = $accessedColumns;
+	}
+
+
+	public function setAccessedColumn(string $key, bool $value): void
+	{
+		$this->accessedColumns[$key] = $value;
+	}
+
+
+	/**
+	 * Loads cache of previous accessed columns and returns it.
+	 * @return array|bool
+	 */
+	public function getPreviousAccessedColumns()
+	{
+		if ($this->cache && $this->previousAccessedColumns === null) {
+			$this->accessedColumns = $this->previousAccessedColumns = $this->cache->load($this->getGeneralCacheKey());
+			if ($this->previousAccessedColumns === null) {
+				$this->previousAccessedColumns = [];
+			}
+		}
+
+		return array_keys(array_filter((array) $this->previousAccessedColumns));
+	}
+
+
+	/**
+	 * @param mixed
+	 */
+	public function setPreviousAccessedColumns($previousAccessedColumns): void
+	{
+		$this->previousAccessedColumns = $previousAccessedColumns;
+	}
+
+
+	public function saveState(): void
+	{
+		if ($this->observeCache === $this->selection && $this->cache && !$this->selection->getSqlBuilder()->getSelect() && $this->accessedColumns !== $this->previousAccessedColumns) {
+			$previousAccessed = $this->cache->load($this->getGeneralCacheKey());
+			$accessed = $this->accessedColumns;
+			$needSave = is_array($accessed) && is_array($previousAccessed)
+				? array_intersect_key($accessed, $previousAccessed) !== $accessed
+				: $accessed !== $previousAccessed;
+
+			if ($needSave) {
+				$save = is_array($accessed) && is_array($previousAccessed) ? $previousAccessed + $accessed : $accessed;
+				$this->cache->save($this->getGeneralCacheKey(), $save);
+				$this->previousAccessedColumns = null;
+			}
+		}
+	}
+
+
+	public function &loadFromRefCache(&$referencing): string
+	{
+		$hash = $this->getSpecificCacheKey();
+		$this->observeCache = &$referencing['observeCache'];
+		$this->accessedColumns = &$referencing[$hash]['accessed'];
+		$this->specificCacheKey = &$referencing[$hash]['specificCacheKey'];
+
+		return $hash;
+	}
+
+
+	/**
+	 * @param mixed
```

Comment:
$observeCache has typehint bool and here is mixed

Reply:
Fixed, allowed only `Selection`.